src/garage/tf/algos/npo2.py

```diff
@@ -0,0 +1,562 @@
+                                          policy_entropy)
+
+        with tf.name_scope('policy_loss'):
+            adv = compute_advantages(self.discount,
+                                     self.gae_lambda,
+                                     self.max_path_length,
+                                     i.baseline_var,
+                                     rewards,
+                                     name='adv')
+
+            adv = tf.reshape(adv, [-1, self.max_path_length])
+            if not self.policy.recurrent:
+                adv = tf.boolean_mask(adv, i.valid_var)
+
+            # Optionally normalize advantages
+            eps = tf.constant(1e-8, dtype=tf.float32)
+            if self.center_adv:
+                adv = center_advs(adv, axes=[0], eps=eps)
+
+            if self.positive_adv:
+                adv = positive_advs(adv, eps)
+
+            with tf.name_scope('kl'):
+                kl = self._old_policy.distribution.kl_divergence(
+                    self.policy.distribution)
+                pol_mean_kl = tf.reduce_mean(kl)
+
+            # Calculate vanilla loss
+            with tf.name_scope('vanilla_loss'):
+                ll = self.policy.distribution.log_prob(i.action_var,
+                                                       name='log_likelihood')
+                vanilla = ll * adv
+
+            # Calculate surrogate loss
+            with tf.name_scope('surrogate_loss'):
+                lr = tf.exp(
+                    ll - self._old_policy.distribution.log_prob(i.action_var))
+                surrogate = lr * adv
+
+            # Finalize objective function
+            with tf.name_scope('loss'):
+                if self._pg_loss == 'vanilla':
+                    # VPG uses the vanilla objective
+                    obj = tf.identity(vanilla, name='vanilla_obj')
+                elif self._pg_loss == 'surrogate':
+                    # TRPO uses the standard surrogate objective
+                    obj = tf.identity(surrogate, name='surr_obj')
+                elif self._pg_loss == 'surrogate_clip':
+                    lr_clip = tf.clip_by_value(lr,
+                                               1 - self._lr_clip_range,
+                                               1 + self._lr_clip_range,
+                                               name='lr_clip')
+                    surr_clip = lr_clip * adv
+                    obj = tf.minimum(surrogate, surr_clip, name='surr_obj')
+
+                if self._entropy_regularzied:
+                    obj += self._policy_ent_coeff * policy_entropy
+
+                # Maximize E[surrogate objective] by minimizing
+                # -E_t[surrogate objective]
+                if self.policy.recurrent:
+                    # filter valid obj
+                    obj = tf.boolean_mask(obj, i.valid_var)
+                loss = -tf.reduce_mean(obj)
+
+            # Diagnostic functions
+            self._f_policy_kl = compile_function(
```

Comment:
can you use tf.Session.make_callable instead?

Reply:
Sure